Feat(AIGW): On-prem config doc (#5866)
Kong/developer.konghq.com: app/ai-gateway/configure-on-prem.md

---
title: "Configure {{site.ai_gateway_name}} on-prem"

description: "Configure {{site.ai_gateway_name}} on self-hosted {{site.base_gateway}} using the 3.x data model and AI plugins, and map each core entity to its plugin."
content_type: reference
layout: reference
products:
  - ai-gateway

works_on:
  - on-prem

breadcrumbs:
  - /ai-gateway/

min_version:
  ai-gateway: '2.0'

related_resources:
  - text: "{{site.ai_gateway}} entities"
    url: /ai-gateway/entities/
  - text: "{{site.ai_gateway}} Policies"
    url: /ai-gateway/policies/
  - text: AI Proxy Advanced plugin
    url: /plugins/ai-proxy-advanced/
  - text: AI MCP Proxy plugin
    url: /plugins/ai-mcp-proxy/
  - text: AI A2A Proxy plugin
    url: /plugins/ai-a2a-proxy/
---

> On {{site.konnect_short_name}}, you configure {{site.ai_gateway}} through its entity model: [AI Providers](/ai-gateway/entities/ai-provider/), [AI Models](/ai-gateway/entities/ai-model/), [AI MCP Servers](/ai-gateway/entities/ai-mcp-server/), and [AI Agents](/ai-gateway/entities/ai-agent/). On self-hosted on-prem {{site.base_gateway}}, you can configure the same capabilities with {{site.ai_gateway}} plugins. You create [Services](/gateway/entities/service/) and [Routes](/gateway/entities/route/) and attach the relevant plugin, rather than declaring entities.

Reviewer: ```suggestion
On {{site.konnect_short_name}}, you configure {{site.ai_gateway}} through its entity model: [AI Model Providers](/ai-gateway/entities/ai-provider/), [AI Identity Providers](/ai-gateway/entities/ai-identity-provider) [AI Models](/ai-gateway/entities/ai-model/), [AI MCP Servers](/ai-gateway/entities/ai-mcp-server/), and [AI Agents](/ai-gateway/entities/ai-agent/). On self-hosted on-prem {{site.base_gateway}}, you can configure the same capabilities with {{site.ai_gateway}} plugins. You create [Services](/gateway/entities/service/) and [Routes](/gateway/entities/route/) and attach the relevant plugin, rather than declaring entities.
```

Reviewer: AI Providers are now split into AI Model Providers (which carry outbound auth settings to LLM services) and AI Identity Providers (which control inbound access to AI Models through key auth and/or OIDC). I am working on this change in another PR

{:.info}

> > On-prem, each plugin's configuration maps 1:1 to an [{{site.ai_gateway}} Policy](/ai-gateway/policies/). The Policy fields and the plugin fields are the same.

Reviewer: ```suggestion
{:.info}
> On-prem, each plugin's configuration maps 1:1 to an [{{site.ai_gateway}} Policy](/ai-gateway/policies/). The AI Policy fields and the plugin fields are the same.
```

## How configuration differs by deployment

> The following table maps each capability to its {{site.konnect_short_name}} entity and its on-prem plugin:

Reviewer: Here's the input from the AI GW runtime team i got for the architecture doc. I think we must be much more detailed in this document than we are currently. Also, I think we have to document Routes created by Agents and MCP Servers and how these entities translate to plugin instances on the Gateway level.

## How entities become {{site.base_gateway}} configuration

AI entities are a high-level abstraction. When you save one, a dedicated conversion step lowers it into the same {{site.base_gateway}} building blocks classic {{site.base_gateway}} uses (Routes, Services, Plugins, Consumers), and that's what data plane nodes actually run.

The mapping isn't uniform. Some entities carry over almost directly: an AI Vault becomes a Kong Vault, and an AI Consumer or AI Consumer Group becomes a Kong Consumer or Consumer Group with its credentials and group membership. An AI Model fans out further: it becomes a Service, one or more Routes, and plugins such as `ai-model-selector` and `ai-proxy-advanced`. An AI Provider has no object of its own. It's folded into the `ai-proxy-advanced` target of every AI Model that references it. AI Policy is the most dynamic case: its `type` becomes the name of whichever {{site.base_gateway}} plugin implements it (for example `ai-prompt-guard` or `ai-rate-limiting-advanced`), applied globally or scoped to whatever the policy is attached to.

### Example: an AI Model becomes {{site.base_gateway}} configuration

An AI Model named `gpt-5-2` with one target routes to an AI Provider `openai-prod`:

```yaml
# AI entities, as configured in {{site.konnect_short_name}}
models:
  - name: gpt-5-2
    capabilities: [generate]
    formats: [{ type: openai }]
    config:
      route: { paths: [/ai] }
      model: { alias: "@openai/gpt-5.2" }
    targets:
      - name: gpt-5.2
        provider: openai-prod
        config: { type: openai, temperature: 1.0, max_tokens: 1024 }
providers:
  - name: openai-prod
    type: openai
    config:
      auth: { type: basic, headers: [{ name: Authorization, value: "{vault://ai/openai-token}" }] }
```

become these {{site.base_gateway}} objects on the data plane:

```yaml
services:
  - name: ai-gateway
    routes:
      - name: openai-chat
        paths: [/ai/chat/completions]
        methods: [POST]
plugins:
  - name: ai-model-selector
    route: { name: openai-chat }
    config: { source: body, body_path: model }
  - name: ai-proxy-advanced
    route: { name: openai-chat }
    model: { name: gpt-5-2 }
    config:
      llm_format: openai
      balancer: { algorithm: round-robin }
      targets:
        - model: { name: gpt-5.2, model_alias: "@openai/gpt-5.2", provider: openai,
                    options: { temperature: 1, max_tokens: 1024 } }
          auth: { header_name: Authorization, header_value: "{vault://ai/openai-token}" }
ai-models:
  - { name: gpt-5-2, alias: "@openai/gpt-5.2" }
```

The AI Provider has no object of its own. Its auth and type are folded directly into the `ai-proxy-advanced` target.

Reviewer: Also this paragraph should be something like 

The following table maps each capability to its {{site.konnect_short_name}} entity, its on-prem plugin, and the Service/Route topology each one needs. See [Models](#models), [MCP Servers](#mcp-servers), and [Agents](#agents) below for plugin configuration details.

> 
> ## How configuration differs by deployment
> 
> The following table maps each capability to its {{site.konnect_short_name}} entity and its on-prem plugin:
> 
> {% table %}
> columns:
>   - title: Capability
>     key: capability
>   - title: "{{site.konnect_short_name}} entity"
>     key: entity
>   - title: On-prem plugin
>     key: plugin
> rows:
>   - capability: Models
>     entity: AI Model
>     plugin: "[`ai-proxy-advanced`](/plugins/ai-proxy-advanced/)"
>   - capability: MCP Servers
>     entity: AI MCP Server
>     plugin: "[`ai-mcp-proxy`](/plugins/ai-mcp-proxy/)"
>   - capability: Agents
>     entity: AI Agent
>     plugin: "[`ai-a2a-proxy`](/plugins/ai-a2a-proxy/)"
> {% endtable %}

Reviewer: ```suggestion
## How configuration differs by deployment

The following table maps each capability to its {{site.konnect_short_name}} entity, its on-prem plugin, and the Service/Route topology each one needs:

{% table %}
columns:
  - title: Capability
    key: capability
  - title: "{{site.konnect_short_name}} entity"
    key: entity
  - title: On-prem plugin
    key: plugin
  - title: On-prem primitives
    key: primitives
rows:
  - capability: Models
    entity: AI Model
    plugin: "[`ai-proxy-advanced`](/plugins/ai-proxy-advanced/)"
    primitives: 1 Service + 1 Route per capability, each with its own plugin instance
  - capability: MCP Servers
    entity: AI MCP Server
    plugin: "[`ai-mcp-proxy`](/plugins/ai-mcp-proxy/)"
    primitives: 1 Service + 1 Route in passthrough or listener mode, or 2 Routes in conversion-listener mode, each with its plugin instance
  - capability: Agents
    entity: AI Agent
    plugin: "[`ai-a2a-proxy`](/plugins/ai-a2a-proxy/)"
    primitives: 1 Service + 1 Route + 1 plugin instance
{% endtable %}

```

## On-prem configuration

On-prem, a client sends requests to {{site.base_gateway}}, where a [Service](/gateway/entities/service/) and [Route](/gateway/entities/route/) carry the AI plugin. The plugin applies the AI behavior and proxies the request to the upstream, whether that's an LLM provider, an MCP server, or an agent.

{% mermaid %}
flowchart LR
  Client(Client)
  subgraph Gateway["{{site.base_gateway}}"]
    P1[AI Proxy Advanced]
    P2[AI MCP Proxy]
    P3[AI A2A Proxy]
  end
  LLM(LLM providers)
  MCP(MCP servers)
  Agent(Agents)
  Client --> P1 --> LLM
  Client --> P2 --> MCP
  Client --> P3 --> Agent
{% endmermaid %}
> _Figure 1:_ On-prem, {{site.ai_gateway}} capabilities are delivered by plugins on {{site.base_gateway}}, each proxying to its upstream.

> ## Models

Reviewer: ```suggestion
## AI Models
```

Reviewer: I think we have to be more detailed here: this is split between the AI Model Provider and AI Model entities

> Use the [AI Proxy Advanced](/plugins/ai-proxy-advanced/) (`ai-proxy-advanced`) plugin to transform and proxy requests to multiple AI providers and models at the same time, and to load balance across targets.

Reviewer: ```suggestion
Use the [AI Proxy Advanced](/plugins/ai-proxy-advanced/) (`ai-proxy-advanced`) plugin to transform and proxy requests to multiple AI providers and models at the same time, and to load balance across targets. On {{site.konnect_short_name}}, an AI Model generates one Route per capability (chat completions, embeddings, and so on), each with its own `ai-proxy-advanced` plugin instance. On-prem, create one Route per capability you want to expose, each with its own `ai-proxy-advanced` plugin.

```

Reviewer: I think we have to be more explicit about how this translates to Konnect

> ## MCP Servers

Reviewer: ```suggestion
## AI MCP Servers
```

> Use the [AI MCP Proxy](/plugins/ai-mcp-proxy/) (`ai-mcp-proxy`) plugin to convert APIs into MCP tools, proxy MCP servers, expose MCP tools to AI clients, and observe MCP traffic.

Reviewer: ```suggestion
Use the [AI MCP Proxy](/plugins/ai-mcp-proxy/) (`ai-mcp-proxy`) plugin to convert APIs into MCP tools, proxy MCP servers, expose MCP tools to AI clients, and observe MCP traffic. On {{site.konnect_short_name}}, an AI MCP Server generates one Route in passthrough or listener mode, or two Routes in conversion-listener mode. On-prem, match that Route count for the mode you're configuring.

```

> ## Agents

Reviewer: ```suggestion
## AI Agents
```

> Use the [AI A2A Proxy](/plugins/ai-a2a-proxy/) (`ai-a2a-proxy`) plugin to add observability and gateway control to Agent-to-Agent (A2A) protocol traffic. The plugin supports both JSON-RPC and REST bindings.

Reviewer: ```suggestion
Use the [AI A2A Proxy](/plugins/ai-a2a-proxy/) (`ai-a2a-proxy`) plugin to add observability and gateway control to Agent-to-Agent (A2A) protocol traffic. The plugin supports both JSON-RPC and REST bindings. On {{site.konnect_short_name}}, an AI Agent generates one Service, one Route, and one `ai-a2a-proxy` plugin, the same 1:1 shape you'd configure on-prem.
```

> ## Consumers, Consumer Groups, and Vaults
> 
> Access control and secret management on-prem use the same {{site.base_gateway}} objects as any other Gateway configuration, so there's no AI-specific setup. Use the existing {{site.base_gateway}} documentation:
> 
> * [Consumers](/gateway/entities/consumer/): Authenticate the clients that call your AI routes.
> * [Consumer Groups](/gateway/entities/consumer-group/): Apply shared rate limits and policies to groups of Consumers.
> * [Vaults](/gateway/entities/vault/): Store and reference provider credentials.

Reviewer: ```suggestion
## Consumers, Consumer Groups, and Vaults

Access control and secret management on-prem use the same {{site.base_gateway}} objects as any other Gateway configuration, so the objects themselves need no AI-specific setup. Use the existing {{site.base_gateway}} documentation:

* [Consumers](/gateway/entities/consumer/): Authenticate the clients that call your AI routes. For Consumer-scoped behavior, such as OpenID Connect authentication, attach the plugin directly to that Consumer.
* [Consumer Groups](/gateway/entities/consumer-group/): Apply shared rate limits and policies to groups of Consumers by attaching the plugin to the Consumer Group.
* [Vaults](/gateway/entities/vault/): Store and reference provider credentials.
```

related_resources:
  - text: "{{site.ai_gateway}} entities"
    url: /ai-gateway/entities/
  - text: "{{site.ai_gateway}} Policies"
    url: /ai-gateway/policies/
  - text: AI Proxy Advanced plugin
    url: /plugins/ai-proxy-advanced/
  - text: AI MCP Proxy plugin
    url: /plugins/ai-mcp-proxy/
  - text: AI A2A Proxy plugin
    url: /plugins/ai-a2a-proxy/
---

On {{site.konnect_short_name}}, you configure {{site.ai_gateway}} through its entity model: [AI Providers](/ai-gateway/entities/ai-provider/), [AI Models](/ai-gateway/entities/ai-model/), [AI MCP Servers](/ai-gateway/entities/ai-mcp-server/), [AI Agents](/ai-gateway/entities/ai-agent/), and [AI Policies](/ai-gateway/entities/ai-policy/). Self-hosted {{site.base_gateway}} doesn't expose these entities. Instead, you configure the same capabilities with AI plugins on [Services](/gateway/entities/service/) and [Routes](/gateway/entities/route/).

Both deployments run the same {{site.base_gateway}} primitives. When you save an AI entity in {{site.konnect_short_name}}, {{site.ai_gateway}} generates the Services, Routes, Plugins, and Consumers that data planes run. On-prem, you create those primitives yourself.

{:.info}

> > On-prem, each plugin's configuration maps 1:1 to an [{{site.ai_gateway}} Policy](/ai-gateway/policies/). The AI Policy fields and the plugin fields are the same.

Reviewer: ```suggestion
On {{site.konnect_short_name}}, you configure {{site.ai_gateway}} through its entity model: [AI Models](/ai-gateway/entities/ai-model/), [AI Model Providers](/ai-gateway/entities/ai-model-provider/), [AI MCP Servers](/ai-gateway/entities/ai-mcp-server/), [AI Agents](/ai-gateway/entities/ai-agent/), [AI Identity Providers](/ai-gateway/entities/ai-identity-provider/), [AI Policies](/ai-gateway/entities/ai-policy/), [AI Consumers](/ai-gateway/entities/ai-consumer/), [AI Consumer Groups](/ai-gateway/entities/ai-consumer-group/), and [AI Vaults](/ai-gateway/entities/ai-vault/). Self-hosted {{site.base_gateway}} doesn't expose these entities. Instead, you configure the same capabilities with AI plugins on [Services](/gateway/entities/service/) and [Routes](/gateway/entities/route/).

Both deployments run the same {{site.base_gateway}} primitives. When you save an AI entity in {{site.konnect_short_name}}, {{site.ai_gateway}} generates the Services, Routes, Plugins, and Consumers that data planes run. On-prem, you create those primitives yourself.

{:.info}
> On-prem, each Policy-backed plugin's configuration maps 1:1 to an [{{site.ai_gateway}} Policy](/ai-gateway/policies/). The AI Policy fields and the plugin fields are the same.
```

## How entities become {{site.base_gateway}} configurations

> The mapping depends on the entity. Some entities generate a single {{site.base_gateway}} primitive, some generate several, and an AI Provider generates none of its own.

Reviewer: ```suggestion
## How entities become {{site.base_gateway}} configuration

AI entities are a high-level abstraction. When you save one, a dedicated conversion step lowers it into the same {{site.base_gateway}} building blocks classic {{site.base_gateway}} uses (Routes, Services, Plugins, Consumers), and that's what data plane nodes actually run.

The mapping isn't uniform. Some entities carry over almost directly: an AI Vault becomes a Kong Vault, and an AI Consumer or AI Consumer Group becomes a Kong Consumer or Consumer Group with its credentials and group membership. An AI Model fans out further: it becomes a Service, one or more Routes, and the AI Proxy Advanced plugin. An AI Provider has no object of its own. It's folded into the AI Proxy Advanced target of every AI Model that references it. AI Policy is the most dynamic case: its type becomes the name of whichever {{site.base_gateway}} plugin implements it (for example AI Prompt Guard or AI Rate Limiting Advanced), applied globally or scoped to whatever the policy is attached to.
```

> {% table %}
> columns:
>   - title: "{{site.konnect_short_name}} entity"
>     key: entity
>   - title: On-prem {{site.base_gateway}} configuration
>     key: primitives
> rows:
>   - entity: "[AI Model](/ai-gateway/entities/ai-model/)"
>     primitives: "A Service, one Route per capability it serves, and the plugins on each Route (`ai-model-selector` and [`ai-proxy-advanced`](/plugins/ai-proxy-advanced/))."
>   - entity: "[AI Provider](/ai-gateway/entities/ai-provider/)"
>     primitives: "None of its own. Its `type` and credentials are materialized into the `ai-proxy-advanced` target of every AI Model that references it."
>   - entity: "[AI MCP Server](/ai-gateway/entities/ai-mcp-server/)"
>     primitives: "One or more Routes carrying the [`ai-mcp-proxy`](/plugins/ai-mcp-proxy/) plugin. The Route topology depends on the server [mode](/ai-gateway/entities/ai-mcp-server/#server-modes)."
>   - entity: "[AI Agent](/ai-gateway/entities/ai-agent/)"
>     primitives: "A Service, a Route, and the [`ai-a2a-proxy`](/plugins/ai-a2a-proxy/) plugin."
>   - entity: "[AI Policy](/ai-gateway/entities/ai-policy/)"
>     primitives: "The {{site.base_gateway}} plugin named by the policy `type` (for example, `ai-prompt-guard` or `ai-rate-limiting-advanced`), applied globally or scoped to whatever the policy is attached to."
>   - entity: "[AI Consumer](/ai-gateway/entities/ai-consumer/)"
>     primitives: "A [Consumer](/gateway/entities/consumer/) with its credentials."
>   - entity: "[AI Consumer Group](/ai-gateway/entities/ai-consumer-group/)"
>     primitives: "A [Consumer Group](/gateway/entities/consumer-group/) with its membership."
>   - entity: "[AI Vault](/ai-gateway/entities/ai-vault/)"
>     primitives: "A [Vault](/gateway/entities/vault/)."
> {% endtable %}

Reviewer: ```suggestion
{% table %}
columns:
  - title: "{{site.konnect_short_name}} entity"
    key: entity
  - title: On-prem {{site.base_gateway}} configuration
    key: primitives
rows:
  - entity: "[AI Model](/ai-gateway/entities/ai-model/)"
    primitives: "A Service, one Route per capability it serves, and the AI Proxy Advanced plugin on each Route."
  - entity: "[AI Provider](/ai-gateway/entities/ai-provider/)"
    primitives: "None of its own. Its `type` and credentials are materialized into the AI Proxy Advanced target of every AI Model that references it."
  - entity: "[AI MCP Server](/ai-gateway/entities/ai-mcp-server/)"
    primitives: "One or more Routes carrying the AI MCP Proxy plugin. The Route topology depends on the server [mode](/ai-gateway/entities/ai-mcp-server/#server-modes)."
  - entity: "[AI Agent](/ai-gateway/entities/ai-agent/)"
    primitives: "A Service, a Route, and the AI A2A Proxy plugin."
  - entity: "[AI Policy](/ai-gateway/entities/ai-policy/)"
    primitives: "The {{site.base_gateway}} plugin named by the policy `type` (for example, AI Prompt Guard or AI Rate Limiting Advanced), applied globally or scoped to whatever the policy is attached to."
  - entity: "[AI Consumer](/ai-gateway/entities/ai-consumer/)"
    primitives: "A [Consumer](/gateway/entities/consumer/) with its credentials."
  - entity: "[AI Consumer Group](/ai-gateway/entities/ai-consumer-group/)"
    primitives: "A [Consumer Group](/gateway/entities/consumer-group/) with its membership."
  - entity: "[AI Vault](/ai-gateway/entities/ai-vault/)"
    primitives: "A [Vault](/gateway/entities/vault/)."
{% endtable %}
```
